pkg/services/ngalert/state/state_tracker.go

```diff
@@ -54,36 +57,36 @@ func NewStateTracker(logger log.Logger) *StateTracker {
 	return tracker
 }
 
-func (st *StateTracker) getOrCreate(alertRule *ngModels.AlertRule, result eval.Result) AlertState {
+func (st *StateTracker) getOrCreate(alertRule *ngModels.AlertRule, result eval.Result, processingTime time.Duration) AlertState {
 	st.stateCache.mu.Lock()
 	defer st.stateCache.mu.Unlock()
-	lbs := data.Labels{}
-	if len(result.Instance) > 0 {
-		lbs = result.Instance
-	}
+
+	// if duplicate labels exist, alertRule label will take precedence
+	lbs := mergeLabels(alertRule.Labels, result.Instance)
 	lbs["__alert_rule_uid__"] = alertRule.UID
 	lbs["__alert_rule_namespace_uid__"] = alertRule.NamespaceUID
-	lbs["__alert_rule_title__"] = alertRule.Title
-
-	annotations := map[string]string{}
-	if len(alertRule.Annotations) > 0 {
-		annotations = alertRule.Annotations
-	}
+	lbs[prometheusModel.AlertNameLabel] = alertRule.Title
 
 	idString := fmt.Sprintf("%s", map[string]string(lbs))
 	if state, ok := st.stateCache.cacheMap[idString]; ok {
 		return state
 	}
 
+	annotations := map[string]string{}
+	if len(alertRule.Annotations) > 0 {
+		annotations = alertRule.Annotations
+	}
+
 	st.Log.Debug("adding new alert state cache entry", "cacheId", idString, "state", result.State.String(), "evaluatedAt", result.EvaluatedAt.String())
 	newState := AlertState{
-		UID:         alertRule.UID,
-		OrgID:       alertRule.OrgID,
-		CacheId:     idString,
-		Labels:      lbs,
-		State:       result.State,
-		Results:     []StateEvaluation{},
-		Annotations: annotations,
+		AlertRuleUID:   alertRule.UID,
+		OrgID:          alertRule.OrgID,
+		CacheId:        idString,
+		Labels:         lbs,
+		State:          result.State,
+		Results:        []StateEvaluation{},
```

Comment:
Should this be empty?

Reply:
No, it makes more sense to create a new slice with one entry just before creating the new instance of State. Fixed.